### PR TITLE
Fix headers

### PR DESCRIPTION
**Pull request by sbrunner**
## Description

This pull request fixes two issues regarding the Request headers in FCGI

First when the default constructor of `QgsServerRequest` is called, the `mRequestHeaderConv` map will not be filled.
Second, the copy constructor of `QgsServerRequest` is called with an instance of `QgsFcgiServerRequest` in argument, then the override of the `header` method will not work anymore, to fix that we fill the `mHeaders` dictionary by using `setHeader` with the known headers (which one are displayed on request reception).

Then:
* Fix the initialization of the `mRequestHeaderConv` dictionary on FCGI request
* Fill the headers on FCGI request
* Change a bit the header display on new request message

Thanks for reading,
If you don't agree that's on the 3.22 release, I will do that only on master.

Happy Christmas :-)

### CODE REVIEW

File: src/server/qgsfcgiserverrequest.cpp [hunk 1]
```diff
@@ -223,41 +223,57 @@ void QgsFcgiServerRequest::printRequestInfos( const QUrl &url )
...
+  QgsMessageLog::logMessage( QStringLiteral( "Headers:" ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  QgsMessageLog::logMessage( QStringLiteral( "------------------------------------------------" ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  for ( const auto &header : headers )
+  {
+    if ( ! this->header( header ).isEmpty() )
+    {
+      QgsMessageLog::logMessage( QStringLiteral( "%1: %2" ).arg( header ).arg( this->header( header ) ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+      // Fill the headers dictionary
+      this->setHeader( header, this->header( header ) );
+    }
```
Comment (m-kuhn) [inline src/server/qgsfcgiserverrequest.cpp:275]:
```suggestion
  for ( const auto &headerName : headers )
  {
    if ( !header( headerName ).isEmpty() )
    {
      QgsMessageLog::logMessage( QStringLiteral( "%1: %2" ).arg( headerName, header( headerName ) ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
      // Fill the headers dictionary
      setHeader( headerName , header( headerName ) );
    }
```

File: src/server/qgsserverrequest.cpp [hunk 2]
```diff
@@ -21,6 +21,11 @@
 #include <QUrlQuery>
 
 
+QgsServerRequest::QgsServerRequest( )
+{
+  this->init();
```
Comment (m-kuhn) [inline src/server/qgsserverrequest.cpp:26]: No need to use `this`

File: src/server/qgsserverrequest.h [hunk 3]
```diff
@@ -260,6 +260,12 @@ class SERVER_EXPORT QgsServerRequest
     void setBaseUrl( const QUrl &url );
 
   private:
+
+    /**
+     * Initialize the mRequestHeaderConv content
```
Comment (m-kuhn) [inline src/server/qgsserverrequest.h:265]: Can you make this a bit more verbose (variable names are an implementation detail, doc strings should be on a functional level)

**Comment by m-kuhn** [comment]
Thanks for the pull request and the updates @sbrunner 

One minor thing, can we please target master first and than backport to 3.22, that's the safe and future proof approach.

@elpaso @rldhont I'd be happy to get your eyes on this.

**Comment by rldhont** [comment]
@sbrunner can you add some tests ?

**Comment by rldhont** [comment]
@dmarteau can you take a look ?

File: src/server/qgsfcgiserverrequest.cpp [hunk 4]
```diff
@@ -106,11 +106,31 @@ QgsFcgiServerRequest::QgsFcgiServerRequest()
...
+    QStringLiteral( "Authorization" ),
+    QStringLiteral( "X-Qgis-Service-Url" ),
+    QStringLiteral( "X-Qgis-WMS-Service-Url" ),
+    QStringLiteral( "X-Qgis-WFS-Service-Url" ),
+    QStringLiteral( "X-Qgis-WCS-Service-Url" ),
+    QStringLiteral( "X-Qgis-WMTS-Service-Url" ),
+    QStringLiteral( "Forwarded" ),
+    QStringLiteral( "X-Forwarded-Host" ),
+    QStringLiteral( "X-Forwarded-Proto" ),
+    QStringLiteral( "Host" )
```
Comment (pblottiere) [inline src/server/qgsfcgiserverrequest.cpp:124]: What do you think about moving this list in the `QgsServerSettingsEnv::EnvVar` enum with the `HTTP_` prefix instead?

Comment (pblottiere) [reply src/server/qgsfcgiserverrequest.cpp:124]: or using/completing those defined in the `QgsServerRequest::RequestHeader` enum?

Comment (sbrunner) [reply src/server/qgsfcgiserverrequest.cpp:124]: @pblottiere what do you think?

Comment (pblottiere) [reply src/server/qgsfcgiserverrequest.cpp:124]: I'm not sure, I'm going to take a look. And yes I was indeed thinking about relying on a `QMetaEnum`.

Comment (sbrunner) [reply src/server/qgsfcgiserverrequest.cpp]: I'm starting on an implementation based on QMetaEnum

File: src/server/qgsfcgiserverrequest.cpp [hunk 5]
```diff
@@ -106,11 +107,19 @@ QgsFcgiServerRequest::QgsFcgiServerRequest()
   setUrl( url );
   setMethod( method );
 
-  // Get accept header for content-type negotiation
-  const char *accept = getenv( "HTTP_ACCEPT" );
-  if ( accept )
+  // Fill the headers dictionary
+  const QMetaEnum headerMetaEnum = QMetaEnum::fromType<QgsServerRequest::RequestHeader>();
+  for ( int idx = 0; idx < headerMetaEnum.keyCount(); ++idx )
```
Comment (3nids) [inline src/server/qgsfcgiserverrequest.cpp:112]: you can use `qgsEnumMap` 
https://qgis.org/api/qgis_8h.html#a496c75b1881ae4913d017446a4858272

Comment (sbrunner) [reply src/server/qgsfcgiserverrequest.cpp]: +1 Thanks :-)

**Comment by sbrunner** [comment]
@rldhont and @pblottiere I updated the pull request with your comment :-)

**Review by rldhont — APPROVED** [review]
Thanks for tests.

**Comment by rldhont** [comment]
@dmarteau do you have a comment ?

File: src/server/qgsfcgiserverrequest.cpp (hunk 4 above)
Comment (pblottiere) [reply src/server/qgsfcgiserverrequest.cpp]: Thanks, we don't have redundant code now thanks to your last modifications :+1:

**Comment by pblottiere** [comment]
Thanks @sbrunner, code looks good to me :+1:. But I don't have much experience with these variables so I'd like to have @elpaso or @dmarteau opitions :).

**Comment by dmarteau** [comment]
These variables matter a lot when the is server behind a proxy.

File: src/server/qgsserverrequest.cpp [hunk 6]
```diff
@@ -32,19 +33,7 @@ QgsServerRequest::QgsServerRequest( const QUrl &url, Method method, const Header
...
-  mRequestHeaderConv.insert( HOST, QStringLiteral( "Host" ) );
-  mRequestHeaderConv.insert( FORWARDED, QStringLiteral( "Forwarded" ) );
-  mRequestHeaderConv.insert( X_FORWARDED_FOR, QStringLiteral( "X-Forwarded-For" ) );
-  mRequestHeaderConv.insert( X_FORWARDED_HOST, QStringLiteral( "X-Forwarded-Host" ) );
-  mRequestHeaderConv.insert( X_FORWARDED_PROTO, QStringLiteral( "X-Forwarded-Proto" ) );
-  mRequestHeaderConv.insert( X_QGIS_SERVICE_URL, QStringLiteral( "X-Qgis-Service-Url" ) );
-  mRequestHeaderConv.insert( X_QGIS_WMS_SERVICE_URL, QStringLiteral( "X-Qgis-Wms-Service-Url" ) );
-  mRequestHeaderConv.insert( X_QGIS_WFS_SERVICE_URL, QStringLiteral( "X-Qgis-Wfs-Service-Url" ) );
-  mRequestHeaderConv.insert( X_QGIS_WCS_SERVICE_URL, QStringLiteral( "X-Qgis-Wcs-Service-Url" ) );
-  mRequestHeaderConv.insert( X_QGIS_WMTS_SERVICE_URL, QStringLiteral( "X-Qgis-Wmts-Service-Url" ) );
```
Comment (dmarteau) [inline src/server/qgsserverrequest.cpp:46]: Is this `QGIS_SERVER_WMTS_SERVICE_URL` that should be tested instead ?

**Review by dmarteau — COMMENTED** [review]
May be I have missed something but I don't understand why there is two way to override service URL's:  `HTTP_X_QGIS_` and `QGIS_SERVER_....`  ?

File: src/server/qgsfcgiserverrequest.cpp [hunk 7]
```diff
@@ -223,41 +230,36 @@ void QgsFcgiServerRequest::printRequestInfos( const QUrl &url )
...
       QgsMessageLog::logMessage( QStringLiteral( "%1: %2" ).arg( envVar ).arg( QString( getenv( envVar.toStdString().c_str() ) ) ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
     }
   }
+
+  QgsMessageLog::logMessage( QStringLiteral( "Headers:" ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  QgsMessageLog::logMessage( QStringLiteral( "------------------------------------------------" ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  for ( const auto &headerName : headers().keys() )
+  {
+    QgsMessageLog::logMessage( QStringLiteral( "%1: %2" ).arg( headerName ).arg( headers().value( headerName ) ), QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
+  }
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply src/server/qgsfcgiserverrequest.cpp:262]: OK, done :-)

**Comment by sbrunner** [comment]
> May be I have missed something but I don't understand why there is two way to override service URL's: HTTP_X_QGIS_ and QGIS_SERVER_.... ?

`QGIS_SERVER_` => static configuration from an environment variable
`X-Qgis-Wmts-Service-Url` => dynamic configuration from a header (=> `HTTP_X_QGIS_` in fcgi)

**Comment by dmarteau** [comment]
> X-Qgis-Wmts-Service-Url => dynamic configuration from a header (=> HTTP_X_QGIS_ in fcgi)

Ok thx,  I did not catch that.    So, what is the précedence between the two configuration ?

**Comment by sbrunner** [comment]
The environment variables :-)

**Comment by sbrunner** [comment]
@dmarteau I just add the missing tests :-)

**Comment by sbrunner** [comment]
OK, I will merge it tomorrow :-)

**Pull request merged**